src/Compilers/Core/Portable/DiagnosticAnalyzer/PerTreeOptionsProvider.cs

```diff
@@ -0,0 +1,90 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.Options;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    /// <summary>
+    /// Provide options keyed on <see cref="SyntaxTree"/>.
+    /// </summary>
+    public abstract class PerTreeOptionsProvider
+    {
+        /// <summary>
+        /// Get options for a given <paramref name="tree"/>.
+        /// Null if there are no options registered for the given
+        /// tree.
+        /// </summary>
+        public abstract OptionSet TryGetOptions(SyntaxTree tree);
+    }
+
+    internal sealed class CompilerPerTreeOptionsProvider : PerTreeOptionsProvider
+    {
+        public static OptionSet EmptyCompilerOptionSet { get; }
+            = new CompilerOptionSet();
+
+        public const string OptionFeatureName = "analyzer-config";
+
+        internal class CompilerOptionSet : OptionSet
```

Comment:
📝 The member layout of the containing type was very difficult to file. Suggest moving the nested class to the end of the type, or separating it to a new file **CompilerPerTreeOptionsProvider+CompilerOptionSet.cs**.

💭 This class is going to demand we fix #27355.